packages/fxa-auth-server/lib/payments/customer-locations.ts

```diff
@@ -0,0 +1,173 @@
+   */
+  mapLongToShortStateName(country: string, state: string): string {
+    if (
+      Object.keys(STATES_LONG_NAME_TO_SHORT_NAME_MAP).includes(country) &&
+      Object.keys(STATES_LONG_NAME_TO_SHORT_NAME_MAP[country]).includes(state)
+    ) {
+      return STATES_LONG_NAME_TO_SHORT_NAME_MAP[country][state];
+    }
+    // We don't care about any other country/state combos.
+    this.log.debug('customerLocations.mapLongToShortStateName', {
+      message: 'State not found in long name to short name map.',
+      state,
+    });
+    return state;
+  }
+
+  /**
+   * Finds the best `country` and `state` location for the user based on a provided
+   * list of inferred locations and the country in their existing Stripe customer
+   * record.
+   *  - If no country is found in Stripe, `state` and `country` are both "unknown".
+   *  - If there is a country mismatch, prefer the country in Stripe and select the most
+   *    frequent location with a matching country.
+   *  - If there is no matching country, `state` is "unknown".
+   */
+  async getBestLocationForPayPalUser(
+    uid: string,
+    results: PayPalUserLocationResult[]
+  ): Promise<PayPalUserLocationResult | null> {
+    let bestLocation = { uid, state: 'unknown', country: 'unknown', count: 1 };
+    // Avoid hitting Stripe 100 rps rate limit in prod by artificially slowing down requests.
+    // After the first runthrough of the script, we should have backfilled everyone in
+    // Firestore, and this delay can be removed.
+    await sleep(200);
+    const country = await this.getPayPalCustomerCountry(uid);
+    if (!country) {
+      this.log.debug('customerLocations.getBestLocationForPayPalUser', {
+        message: 'No country found in Stripe for the user.',
+        uid,
+      });
+      return bestLocation;
+    }
+    if (!this.isCountryNeeded(country)) {
+      this.log.debug('customerLocations.getBestLocationForPayPalUser', {
+        message: `Region information for user's country is not needed.`,
+        uid,
+        country,
+      });
+      return null;
+    }
+    let locations = results.filter((result) => result.uid === uid);
+    if (locations.length === 0) {
+      return { ...bestLocation, country };
+    }
+    // We have one or more locations for this user; get mode location.
+    locations.sort((a, b) => b.count - a.count);
+    locations = locations.map((location) => ({
+      ...location,
+      country: this.mapLongToShortCountryName(location.country),
+      state: this.mapLongToShortStateName(location.country, location.state),
+    }));
+    const modeLocation = locations[0];
+    // Prioritize country from Stripe customer
+    if (modeLocation.country === country) {
+      bestLocation = { ...modeLocation, count: 1 };
+    } else if (locations.some((location) => location.country === country)) {
+      bestLocation = {
+        ...locations.find((location) => location.country === country)!,
+        count: 1,
+      };
```

Comment:
Seems unnecessary to call `some` and `find` since `find` can tell you if some such element exists in the list.